envoy/auth_server/proto/gateway_endpoint.proto

```diff
@@ -0,0 +1,122 @@
+  bool delete = 3;
+}
+
+// A GatewayEndpoint represents a user account's endpoint, which has two primary functions:
+//  1. Identifying which endpoint is being used to make a service request.
+//  2. Allowing configuration of endpoint-specific settings, such as API key authorization, etc.
+//
+// A GatewayEndpoint is associated to a single UserAccount.
+// A single UserAccount can have multiple GatewayEndpoints.
+// Settings related to service requests, such as enforcing API key authorization, are configured per GatewayEndpoint.
+message GatewayEndpoint {
+  // The unique identifier for a UserAccount's endpoint.
+  // It must be passed as the last path segment of the service request URL; eg `/v1/{endpoint_id}`.
+  // This is used to identify the GatewayEndpoint when making a service request.
+  string endpoint_id = 1;
+
+  // The authorization settings for the GatewayEndpoint.
+  Auth auth = 2;
+
+  // The rate limiting settings for the GatewayEndpoint.
+  // This includes both throughput (TPS) limit and the capacity (longer period) limit.
+  RateLimiting rate_limiting = 3;
+
+  // Optional metadata for the GatewayEndpoint, which can be set to any additional information.
+  map<string, string> metadata = 4;
+}
+
+// The authorization settings for a GatewayEndpoint.
+message Auth {
+  enum AuthType {
+    NO_AUTH = 0;
+    API_KEY_AUTH = 1;
+    JWT_AUTH = 2;
+  }
+
+  // The type of authentication being used.
+  AuthType auth_type = 1;
+
+  oneof auth_type_details {
+    Empty no_auth = 2;
+    // The API key authorization settings for the GatewayEndpoint.
+    string api_key = 3;
+    // The JWT authorization settings for the GatewayEndpoint.
+    JWT jwt = 4;
+  }
+}
+
+message JWT {
+  // A map of ProviderUserIDs authorized to access this UserAccount's GatewayEndpoints.
+  map<string, Empty> authorized_users = 2;
+}
+
+// The rate limiting settings for a GatewayEndpoint.
+message RateLimiting {
+  // ThroughputLimit refers to rate limiting per-second (TPS).
+  // This is used to prevent DoS or DDoS attacks, as well as enforce pricing plan limits.
+  int32 throughput_limit = 1;
+  // CapacityLimit refers to rate limiting over longer periods, such as a day, week or month.
+  // This is to prevent abuse of the services provided, as well enforce pricing plan limits.
+  int32 capacity_limit = 2;
+  // The period over which the CapacityLimit is enforced. One of `daily`, `weekly` or `monthly`.
+  CapacityLimitPeriod capacity_limit_period = 3;
+}
+
+// CapacityLimitPeriod is the period over which the capacity limit is enforced.
+// For example: CapacityLimit=`100,000` and CapacityLimitPeriod=`daily`
+// enforces a rate limit of 100,000 requests per day.
+enum CapacityLimitPeriod {
+    CAPACITY_LIMIT_PERIOD_UNSPECIFIED = 0;
+    CAPACITY_LIMIT_PERIOD_DAILY = 1;
+    CAPACITY_LIMIT_PERIOD_WEEKLY = 2;
+    CAPACITY_LIMIT_PERIOD_MONTHLY = 3;
+  }
+
+// An Empty message is used to indicate that a field is not set.
+message Empty {}
```

Comment:
Thoughts on changing this to `NoAuth` so it's very explicit and forward compatible with any weird stuff we'll do?

Please put it next to the other auth wrappers.

Reply:
This is used for both the empty value in the JWT authorized users map and the empty value for `AuthTypeDetails` when no auth is specified. 

IMO `Empty` makes sense here but also with adding an additional `message NoAuth {}` specifically for that fields (admittedly it feels a bit redundant but happy to do either one).

Reply:
This is a better pattern w.r.t longer-term code maintenance so lets do it.

It prevents others from reusing something they shouldn't and then lots of legacy + techdebt before you know it.